Autopilot/Src/gps.cpp

```diff
@@ -0,0 +1,386 @@
+
+    gpsUART->registerDMAReceiveCallback(handle_DMA_input);
+    //GPS Configuration
+    gpsUART->transmit((uint8_t*) PUBX_CONFIG_NMEA, sizeof(PUBX_CONFIG_NMEA));
+    //Turns off receivers (resets them)
+    gpsUART->transmit((uint8_t*) PUBX_SET_GGA_OFF, sizeof(PUBX_SET_GGA_OFF));
+    gpsUART->transmit((uint8_t*) PUBX_SET_VTG_OFF, sizeof(PUBX_SET_VTG_OFF));
+    //Turns on receivers
+    gpsUART->transmit((uint8_t*) PUBX_SET_GGA, sizeof(PUBX_SET_GGA));
+    gpsUART->transmit((uint8_t*) PUBX_SET_VTG, sizeof(PUBX_SET_VTG));
+    
+    isGPSConfigured = true;
+}
+
+void handle_DMA_input() {
+    if (!isGPSConfigured) {
+        //Assumed the wrapper will handle clearing the registers and all.
+        return;
+    }
+
+    bool currentlyParsing = false;
+    uint16_t bufferIndex = 0;
+    uint8_t data = 0;
+
+    std::deque<uint8_t> dma_data = gpsUART->get_rx_queue(); //Copys double eneded queue in UARTPort class that contains data in the DMA registers
+
+    for (int i = 0; i < sizeof(dma_data) && newGGAData == false && newVTGData == false; i++) { // If new data is found, then the loop is terminated 
+        data = dma_data[i];
+
+        if (data == '$') {
+            currentlyParsing = true;
+            bufferIndex = 0;
+        } else if (data == '\r') {
+            if (strncmp(GGA_MESSAGE, generalBuffer, 5 == 0)) {
+                memcpy(ggaValues, generalBuffer, GPS_UART_BUFFER_SIZE);
+                newGGAData = true;
+            } else if (strncmp(VTG_MESSAGE, generalBuffer, 5 == 0)) {
+                memcpy(vtgValues, generalBuffer, GPS_UART_BUFFER_SIZE);
+                newVTGData = true;
+            } 
+        } else if (currentlyParsing) {
+            generalBuffer[bufferIndex] = data;
+            bufferIndex = (bufferIndex + 1) % GPS_UART_BUFFER_SIZE;
+        }
+    }
+}
+
+bool NEOM8::is_check_sum_valid(char *input) {
+    uint16_t index = 0;
+    uint8_t checksum = 0;
+
+    while(input[index] != '*') {
+        checksum ^= input[index];
+        index++;
+    }
+
+    index++;
+    return uint8_to_hex((checksum & 0xF0) >> 4) == input[index] && uint8_to_hex(checksum & 0x0F) == input[index+1];
+}
+
+uint8_t NEOM8::uint8_to_hex(uint8_t toConvert) {
+    uint8_t intConverted = 0;
+
+    if (toConvert >= 0 && toConvert <= 9) {
+        intConverted = toConvert + 0x30;
+    } else if (toConvert >= 0xA && toConvert <= 0xF) {
+        intConverted = toConvert + 0x37;
+    }
+    
+    return intConverted;
+}
+
+int NEOM8::ascii_to_hex(int toConvert) {
+    int intConverted = 0;
+
+    if (toConvert == 46) {
+        intConverted = 46;
+    } else if (toConvert >= 48 && toConvert <= 57) {
+        intConverted = toConvert - 48;
+    } else if (toConvert >= 65 && toConvert <= 70) {
+        intConverted = toConvert - 0x37;
+    }
+
+    return intConverted;
+}
+
+//Checks if there is new data, validates it, and then converts it to a form that the autopilot will understand
+void NEOM8::get_gps_data() {
```

Comment:
I think you forgot to actually return the data lol.

Reply:
I don't need to return the data from this function because in the parse_register() methods it stores the values in private variables of the NEOM8 instance. 

I must've deleted the GetResult() method by accident. Dk when it happened but it did :((